### PR TITLE
Fix serving samples to run as non-root

### PR DESCRIPTION
**Pull request by prushh**
Fixes knative/serving#14566

## Proposed Changes

Changes discussed on #5758

### Overview

- [x] cloudevents-dotnet  -- OK
- [x] cloudevents-go  -- OK
- [x] cloudevents-nodejs  -- OK
- [ ] cloud events-rust -- Need help, build error on [L18](https://github.com/prushh/docs/blob/92854d117e504783c1e4f932f5892dd1c5b8f305/code-samples/serving/cloudevents/cloudevents-rust/Dockerfile#L18) (I wasn't able to create the build locally and I tried to do that using multi-stage container)

```text
RUN rustup target add x86_64-unknown-linux-musl && cargo build --target x86_64-unknown-linux-musl --release:
0.370    info: downloading component 'rust-std' for 'x86_64-unknown-linux-musl'
5.900    info: installing component 'rust-std' for 'x86_64-unknown-linux-musl'
7.002    error: failed to parse lock file at: /Cargo.lock
7.002    Caused by:
7.002    invalid serialized PackageId for key `package.dependencies`
```

- [x] cloudevents-spring  -- No Dockerfile (`mvn compile jib:build -Dimage=<image_name>`)
- [x] cloudevents-vertx  -- No Dockerfile (same above)
- [x] gitwebhook-go  -- OK
- [x] grpc-ping-go  -- OK (`nonroot` tag specified on distroless image)
- [x] helloworld-csharp  -- OK
- [x] helloworld-go  -- OK
- [x] helloworld-java-spark  -- OK
- [x] helloworld-java-spring  -- OK
- [x] helloworld-kotlin  -- OK
- [x] helloworld-nodejs  -- OK
- [x] helloworld-php  -- OK (I'm not sure if it is the correct way to proceed)
- [x] helloworld-python  -- OK
- [x] helloworld-ruby  -- OK
- [x] helloworld-scala  -- Added non-root user, (curl: (52) Empty reply from server)
- [ ] helloworld-shell  -- Need help, incorrect response

  ```html
  <HTML><HEAD><TITLE>404 Not Found</TITLE></HEAD>
  <BODY><H1>404 Not Found</H1>
  The requested URL was not found
  </BODY></HTML>
  ```

- [x] knative-routing-go  -- OK (`nonroot` tag specified on distroless image)
- [x] kong-routing-go  -- OK (`nonroot` tag specified on distroless image)
- [x] servingcontainer  -- OK (bump golang to 1.21, fixed `undefined: io.ReadAll` error)
- [x] sidecarcontainer  -- OK (bump golang to 1.21)
- [x] secrets-go  -- OK

### Additional info

Wherever possible, projects were tested with Docker as follows:

```bash
docker buildx build --platform linux/amd64 -t <image> .
docker run -p 8080:8080 <image>

curl -X POST \
  -H "content-type: application/json"  \
  -H "ce-specversion: 1.0"  \
  -H "ce-source: curl-command"  \
  -H "ce-type: curl.demo"  \
  -H "ce-id: 123-abc"  \
  -d '{"name":"Dave"}' \
  localhost:8080
# OR
curl localhost:8080
```

---

Can you please take a look @ReToCode @kauana?

### CODE REVIEW

File: code-samples/serving/cloudevents/cloudevents-nodejs/Dockerfile [hunk 1]
```diff
@@ -12,27 +12,42 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM registry.access.redhat.com/ubi8/nodejs-12
+FROM node:20-alpine as builder
```
Comment (ReToCode) [inline code-samples/serving/cloudevents/cloudevents-nodejs/Dockerfile:15]: @matzew ok with this change?

File: code-samples/serving/cloudevents/cloudevents-rust/Dockerfile [hunk 2]
```diff
@@ -11,10 +11,29 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+FROM rust:1.31.0 AS builder
```
Comment (ReToCode) [inline code-samples/serving/cloudevents/cloudevents-rust/Dockerfile:14]: probably worth to take a look at https://github.com/knative/docs/pull/5757 as well.

File: code-samples/serving/gitwebhook-go/Dockerfile [hunk 3]
```diff
@@ -32,7 +32,23 @@ RUN CGO_ENABLED=0 GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -o /go/bin/webh
```
Comment (ReToCode) [inline code-samples/serving/gitwebhook-go/Dockerfile:32]: The GH dependency is probably now hosted here: https://github.com/google/go-github

File: code-samples/serving/multi-container/servingcontainer/Dockerfile [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Use the official Golang image to create a build artifact.
 # This is based on Debian and sets the GOPATH to /go.
 # https://hub.docker.com/_/golang
-FROM golang:1.15 as builder
+FROM golang:1.16 as builder
```
Comment (ReToCode) [inline code-samples/serving/multi-container/servingcontainer/Dockerfile:4]: you probably can even go up to 1.21

File: code-samples/serving/multi-container/sidecarcontainer/Dockerfile [hunk 5]
```diff
@@ -1,7 +1,7 @@
 # Use the official Golang image to create a build artifact.
 # This is based on Debian and sets the GOPATH to /go.
 # https://hub.docker.com/_/golang
-FROM golang:1.15 as builder
+FROM golang:1.16 as builder
```
Comment (ReToCode) [inline code-samples/serving/multi-container/sidecarcontainer/Dockerfile:4]: you probably can even go up to 1.21

File: code-samples/serving/secrets-go/Dockerfile [hunk 6]
```diff
@@ -17,11 +17,22 @@ RUN CGO_ENABLED=0 GOOS=linux GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -v -
 # https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds
 FROM alpine
```
Comment (ReToCode) [inline code-samples/serving/secrets-go/Dockerfile:19]: you probably can do the same as with the Serving images:

- add a go.mod: https://github.com/knative/docs/blob/main/code-samples/serving/multi-container/sidecarcontainer/go.mod
- copy and run go mod: https://github.com/knative/docs/blob/main/code-samples/serving/multi-container/sidecarcontainer/Dockerfile#L14

**Review by ReToCode — COMMENTED** [review]
Thanks for doing this @prushh. Seems like lot of images are pretty outdated.
I commented inline, hope that helps with the build errors.

File: code-samples/serving/hello-world/helloworld-shell/Dockerfile [hunk 7]
```diff
@@ -2,9 +2,6 @@
 # https://git.busybox.net/busybox/tree/networking/httpd.c
 FROM busybox
 
-# Serve from this directory
-WORKDIR /var/www
```
Comment (ReToCode) [inline code-samples/serving/hello-world/helloworld-shell/Dockerfile:6]: is this the default?

**Comment by prushh** [comment]
@ReToCode I will work on the missing fixes as soon as possible 😄

File: code-samples/serving/hello-world/helloworld-shell/Dockerfile (hunk 7 above)
Comment (prushh) [reply code-samples/serving/hello-world/helloworld-shell/Dockerfile:6]: Yes, see [L6](https://github.com/knative/docs/blob/62e6f14ff3dce6c5e1ea5b3e30f705dde599fb1e/code-samples/serving/hello-world/helloworld-shell/Dockerfile#L6) on main branch

File: code-samples/serving/secrets-go/Dockerfile (hunk 6 above)
Comment (prushh) [reply code-samples/serving/secrets-go/Dockerfile:19]: I fix the build error with the following steps:

- `go mod init github.com/knative/docs/code-samples/serving/secrets-go`
- `go mod tidy`
- update Dockerfile as you suggest (copy and run go mod)

Is it okay to upload also the _go.sum_ file?

Comment (ReToCode) [reply code-samples/serving/secrets-go/Dockerfile:19]: yes, you can add that as well 👍

Comment (prushh) [reply code-samples/serving/secrets-go/Dockerfile:19]: I choose to execute the commands above in the first stage of the Dockerfile, no more build error and the server starts!

File: code-samples/serving/gitwebhook-go/Dockerfile (hunk 3 above)
Comment (prushh) [reply code-samples/serving/gitwebhook-go/Dockerfile:32]: Same fixes applied to _secrets-go_ project, no more build error and the server starts!

File: code-samples/serving/cloudevents/cloudevents-rust/Dockerfile (hunk 2 above)
Comment (prushh) [reply code-samples/serving/cloudevents/cloudevents-rust/Dockerfile:14]: Sorry for the late reply!
I'm following the PR above as you suggested and the build was successful, but I get an image size of 2.32GB. I also tried to run and curl it but I got `curl: (56) Recv failure: Connection reset by peer` error.

```Dockerfile
FROM rust:1.73.0 AS builder

ARG USER=appuser
ARG USER_UID=1001
ARG USER_GID=$USER_UID

# Add a user so the server will run as a non-root user.
RUN addgroup --gid $USER_GID $USER && \
    adduser -u $USER_UID --ingroup $USER --disabled-password $USER

COPY . .

RUN cargo install --path .

USER $USER

CMD ["knative-cloudevents-example"]
```

Now as a first step I'm trying to reduce the image size by using a second step where I copy and execute the release.

**Comment by ReToCode** [comment]
Hey @prushh, are you still working on the last fixes?

**Comment by prushh** [comment]
Hey @ReToCode !
Unfortunately not, I am currently busy with university... I don't think I'll be able to work on it before mid-March 😪

**Comment by ReToCode** [comment]
@prushh how are things? Are you still willing to work on this PR?

**Comment by ReToCode** [comment]
Yes that is fine, could you please rebase and create the issue with what you found so far?

**Comment by prushh** [comment]
Hi @ReToCode! It was my first rebase, I hope I've done everything correctly.
In the afternoon I should be able to create the issue.

**Comment by BobyMCbobs** [comment]
Let me know whether I can assist with anything!

**Comment by ReToCode** [comment]
@BobyMCbobs I think the work is up for grabs. The PR contains what is done so far, the description should point out the things that are not done yet/not working yet.

**Pull request closed without merging**